libraries/pybamm-developer/03_spm.md

```diff
@@ -0,0 +1,373 @@
+
+$$
+\frac{\partial c_i}{\partial t} = \nabla \cdot (D_i \nabla c_i)
+$$
+
+subject to the following boundary and initial conditions:
+
+$$
+\left.\frac{\partial c_i}{\partial r}\right\vert_{r=0} = 0, \quad \left.\frac{\partial c}{\partial r}\right\vert_{r=R_i} = -j_i, \quad \left.c\right\vert_{t=0} = c^0_i
+$$
+
+where $c_i$ is the concentration of lithium ions in the positive ($i=n$) or
+negative ($i=p$) electrode, $D_i$ is the diffusion coefficient, $j_i$ is the
+interfacial current density, and $c^0_i$ is the concentration at the particle
+surface.
+
+The primary addition here in comparison with our previous PDE model is the
+fluxes of lithium ions in the positive and negative electrodes $j_i$. These are
+dependent on the applied current $I$:
+
+$$
+\begin{align*}
+j_n &= \frac{I}{a_n \delta_n F \mathcal{A}}, \qquad
+j_p &= \frac{-I}{a_p \delta_p F \mathcal{A}},
+\end{align*}
+$$
+
+
+where $a_i = 3 \epsilon_i / R_i$ is the specific surface area of the electrode,
+$\epsilon_i$ is the volume fraction of active material, $\delta_i$ is the
+thickness of the electrode, $F$ is the Faraday constant, and
+$\mathcal{A}$ is the electrode surface area.
+
+## Function Parameters in PyBaMM
+
+The applied current $I(t)$ is a parameter to the model, but unlike the
+other parameters we have seen so far, it is a function of time. We can define
+this using `pybamm.FunctionParameter`:
+
+```python
+I = pybamm.FunctionParameter("Applied current [A]", {"Time [s]": pybamm.t})
+```
+
+This will allow us to define the applied current as a function of time when we
+create our `pybamm.ParameterValues` object.
+
+```python
+def current(t):
+    return 1
+param = pybamm.ParameterValues(
+    {
+        "Applied current [A]": current
+    }
+)
+```
+
+## Domains in PyBaMM
+
+In the SPM model we have two different spatial domains, one for the positive
+electrode and one for the negative electrode. Remember that when we define a variable in
+PyBaMM, we can specify the domain using the `domain` keyword argument:
+
+```python
+c_n = pybamm.Variable("Concentration in negative electrode [mol.m-3]", domain="negative particle")
+```
+
+
+::::challenge{id=spm-state-equations, title=SPM governing equations}
+
+Copy your PDE model from the previous challenge to a new file, and modify it to
+include the state equations for the concentration of lithium ions in both the
+positive and negative electrodes. Note that now you will have two variables,
+$c_n$ and $c_p$, that must be defined on separate spatial domains, and a number
+of new parameters. Define the applied current $I$ as a input parameter that is a
+function of time using `pybamm.FunctionParameter`.
+
+:::solution
+```python
+import pybamm
+
+model = pybamm.BaseModel()
+
+# define parameters
+domains = ["negative particle", "positive particle"]
+I = pybamm.FunctionParameter("Applied current [A]", {"Time [s]": pybamm.t})
+D_i = [pybamm.Parameter(f"Diffusion coefficient for {d} [m2.s-1]") for d in domains]
```

Comment:
This is nice and compact, but I wonder if it ends up being more confusing. Especially when it comes to working with the existing models in pybamm where you couldn't do this because of the way the parameters are named/accessed.

People can't take this example and use it with the parameter sets it PyBaMM because the parameter names do not match.

Reply:
Why not do e.g.
```python3
electrodes = ["negative", "positive"]
D_i = [pybamm.Parameter(f{electrode.capitalize()} electrode diffusivity [m2.s-1]) for electrode in electrodes]
```